Kv spec closes Cromwell #1300

katevoss opened the pull request.

katevoss commented:
@knoblett Could you take a look at my changes to the WDL spec? It's for [issue #1300](https://github.com/broadinstitute/cromwell/issues/1300).

knoblett commented:
Do you still need my review for this? Must've missed this in my inbox earlier, sorry for the delay.

katevoss commented:
@knoblett A quick overview would be great, thank you!

knoblett commented on SPEC.md:
```diff
@@ -2460,6 +2604,8 @@ Would produce the command `python script.py --bams=/path/to/1.bam,/path/to/2.bam
 
 ##### Array serialization using write_lines()
 
+:pig2: [Cromwell supported](https://github.com/broadinstitute/cromwell#wdl-support) :white_check_mark:
```
I am an advocate of specifically also pointing out parts that Cromwell does not support. Something like "planned for a future version of Cromwell" to tell the user that this is not currently functional. Alternatively, we could remove any sections that are not currently implemented. If we go with the latter option, these Cromwell supported bits can be removed, since anything in the document should be supported.

The reason why I point this out is if I am linked to a specific part of the spec, and I don't read other parts, a lack of `Cromwell supported` checkbox isn't very noticeable.

katevoss replied:
@geoffjentry How do you feel about specifically calling out items in the WDL spec that Cromwell does not support? 

🐖  Cromwell not supported ❌  
or
🐖  coming soon in Cromwell

geoffjentry replied:
Out of the two I'd prefer the latter, even though in some cases `soon` might be relative :)

Ultimately I'd like to wind up in a world where the spec doesn't contain bits not supported by cromwell (e.g. moving that kind of stuff to a "proposed future changes" or something like that) but we're not there yet.

knoblett commented on grammar.hgr:
```diff
@@ -3,7 +3,7 @@ grammar {
     partials {
       r'[a-zA-Z]' -> _identifier_start
       r'[a-zA-Z0-9_]' -> _identifier_follow
-      r'(Array|Map|Object|Pair|Boolean|Int|Float|Uri|File|String)(?!{%_identifier_follow%})' -> _type
```
Why was Pair removed here? I thought Pair was recently implemented, though I could have that backwards.

katevoss replied:
I believe edits on the grammar.hgr were unintentional (as a result of my misuse of Github). I will attempt to fix it.

katevoss commented on SPEC.md:
```diff
@@ -1292,6 +1362,8 @@ Conditionals only execute the body if the expression evaluates to true
 
 ### Parameter Metadata
 
+:pig2: Coming soon in [Cromwell](https://github.com/broadinstitute/cromwell)
```
@geoffjentry This is not listed in the spreadsheet, is it in Cromwell or not?

geoffjentry replied:
It exists

katevoss commented on SPEC.md:
```diff
@@ -1312,6 +1384,8 @@ As an example:
 
 ### Metadata
 
+:pig2: Coming soon in [Cromwell](https://github.com/broadinstitute/cromwell)
```
@geoffjentry This is not listed in the spreadsheet, is it in Cromwell or not?

geoffjentry replied:
Also exists

katevoss commented on SPEC.md:
```diff
@@ -2274,9 +2414,23 @@ Default unit is Bytes ("B").
 Given 3 String parameters `input`, `pattern`, `replace`, this function will replace any occurrence matching `pattern` in `input` by `replace`.
 `pattern` is expected to be a [regular expression](https://en.wikipedia.org/wiki/Regular_expression). Details of regex evaluation will depend on the execution engine running the WDL.
 
+## Array[Int] range(Int)
```
@geoffjentry This section wasn't in the spreadsheet, is it supported by Cromwell? I'm not sure why it shows that I added this section, Array[Int] range(Int) and the next few, I think it was a Github error. Does it belong in the spec? 
@ruchim did you add these sections? I might have adopted your changes and now it looks like I made them. Are they supported by Cromwell?

geoffjentry replied:
Range exists

katevoss replied:
@geoffjentry does `Array[Int] range(Int)`, `Array[Array[X]] transpose(Array[Array[X]])`, `Pair(X,Y) zip(X,Y)`, and `Pair(X,Y) cross(X,Y)` exist in Cromwell?

geoffjentry replied:
@katevoss Yep.

katevoss commented:
I think I fixed everything and I added an explicit "Coming soon with Cromwell" note for all 11 WDL features not currently supported.

@knoblett and @geoffjentry  If you could take a quick look over it and make sure I didn't miss anything, that would be great.

geoffjentry commented:
👍

knoblett commented:
Feel free to squash your commits and merge! Im not sure how to do the fancy graphic @geoffjentry did, but pull approved!
